server/cluster_worker.go

```diff
@@ -0,0 +1,414 @@
+package server
+
+import (
+	"math"
+	"math/rand"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pd_jobpb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raft_cmdpb"
+	"github.com/pingcap/kvproto/pkg/raft_serverpb"
+	"github.com/pingcap/kvproto/pkg/raftpb"
+	"github.com/twinj/uuid"
+	"golang.org/x/net/context"
+)
+
+const (
+	checkJobInterval = 10 * time.Second
+
+	connectTimeout = 3 * time.Second
+	readTimeout    = 3 * time.Second
+	writeTimeout   = 3 * time.Second
+)
+
+func (c *raftCluster) onJobWorker() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(checkJobInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.quitCh:
+			return
+		case <-c.askJobCh:
+			if !c.s.IsLeader() {
+				log.Warnf("we are not leader, no need to handle job")
+				continue
+			}
+
+			job, err := c.getJob()
+			if err != nil {
+				log.Errorf("get first job err %v", err)
+			} else if job == nil {
+				// no job now, wait
+				continue
+			}
+			if err = c.handleJob(job); err != nil {
+				log.Errorf("handle job %v err %v, retry", job, err)
+				// wait and force retry
+				time.Sleep(time.Second)
+				asyncNotify(c.askJobCh)
+				continue
+			}
+
+			if err = c.popJob(job); err != nil {
+				log.Errorf("pop job %v err %v", job, err)
+			}
+
+			// Notify to job again.
+			asyncNotify(c.askJobCh)
+		case <-ticker.C:
+			// Try to check job regularly.
+			asyncNotify(c.askJobCh)
+		}
+	}
+}
+
+func asyncNotify(ch chan struct{}) {
+	select {
+	case ch <- struct{}{}:
+	default:
+	}
+}
+
+func (c *raftCluster) postJob(req *raft_cmdpb.RaftCommandRequest) error {
+	jobID, err := c.s.idAlloc.Alloc()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	req.Header.Uuid = uuid.NewV4().Bytes()
+
+	job := &pd_jobpd.Job{
+		JobId:   proto.Uint64(jobID),
+		Status:  pd_jobpd.JobStatus_Pending.Enum(),
+		Request: req,
+	}
+
+	jobValue, err := proto.Marshal(job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	jobPath := makeJobKey(c.clusterRoot, jobID)
+
+	resp, err := c.s.client.Txn(context.TODO()).
+		If(c.s.leaderCmp()).
```

Comment:
TODO, now we miss many checks with etcd. 
I will learn the etcd v3 deeply later and find the best way to communicate with it.